### PR TITLE
Windup 878 - Validating XMLs by it's XSD

### PR DESCRIPTION
**Pull request by mbriskar**
I suggest reviewing by commits. 1st commit does the whole work, second just creates impl/api modules in the rules-xml addon and third improves tests

### CODE REVIEW

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-core-PR-dependents-check/722/

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-pr-builder/1350/

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-pr-builder/1351/

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-core-PR-dependents-check/723/

**Comment by mbriskar** [comment]
Need to add new test testing a valid xml.  Need to test how it works without internet connection.

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-pr-builder/1359/<h2>Failed Tests: <span class='status-failure'>7</span></h2><h3><a name='windup-pr-builder/org.jboss.windup.tests:windup-tests' /><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1359/org.jboss.windup.tests$windup-tests/testReport'>windup-pr-builder/org.jboss.windup.tests:windup-tests</a>: <span class='status-failure'>7</span></h3><ul><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1359/org.jboss.windup.tests$windup-tests/testReport/org.jboss.windup.tests.application/WindupArchitectureMediumBinaryModeTest/testRunWindupMediumWithFernflower/'><strong>org.jboss.windup.tests.application.WindupArchitectureMediumBinaryModeTest.testRunWindupMediumWithFernflower</strong></a></li><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1359/org.jboss.windup.tests$windup-tests/testReport/org.jboss.windup.tests.application/WindupArchitectureMediumBinaryModeTest/testRunWindupMediumWithProcyon/'><strong>org.jboss.windup.tests.application.WindupArchitectureMediumBinaryModeTest.testRunWindupMediumWithProcyon</strong></a></li><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1359/org.jboss.windup.tests$windup-tests/testReport/org.jboss.windup.tests.application/WindupArchitectureScanPackagesTest/testRunWindupScanPackages/'><strong>org.jboss.windup.tests.application.WindupArchitectureScanPackagesTest.testRunWindupScanPackages</strong></a></li><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1359/org.jboss.windup.tests$windup-tests/testReport/org.jboss.windup.tests.application/WindupArchitectureSourceWithBadPOMTest/testRunWindupSourceMode/'><strong>org.jboss.windup.tests.application.WindupArchitectureSourceWithBadPOMTest.testRunWindupSourceMode</strong></a></li><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1359/org.jboss.windup.tests$windup-tests/testReport/org.jboss.windup.tests.application/WindupArchitectureTestKeepArchivesAndGraphTest/testRunWindupDiscard/'><strong>org.jboss.windup.tests.application.WindupArchitectureTestKeepArchivesAndGraphTest.testRunWindupDiscard</strong></a></li><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1359/org.jboss.windup.tests$windup-tests/testReport/org.jboss.windup.tests.application/WindupArchitectureTestKeepArchivesAndGraphTest/testRunWindupKeep/'><strong>org.jboss.windup.tests.application.WindupArchitectureTestKeepArchivesAndGraphTest.testRunWindupKeep</strong></a></li><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1359/org.jboss.windup.tests$windup-tests/testReport/org.jboss.windup.tests.application/WindupMultiAppBasicTest/testRunWindupMedium/'><strong>org.jboss.windup.tests.application.WindupMultiAppBasicTest.testRunWindupMedium</strong></a></li></ul>

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-core-PR-dependents-check/731/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='windup-core-PR-dependents-check/org.jboss.windup.rules:windup-rulesets' /><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-core-PR-dependents-check/731/org.jboss.windup.rules$windup-rulesets/testReport'>windup-core-PR-dependents-check/org.jboss.windup.rules:windup-rulesets</a>: <span class='status-failure'>1</span></h3><ul><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-core-PR-dependents-check/731/org.jboss.windup.rules$windup-rulesets/testReport/org.jboss.windup.rules.tests/WindupRulesTest/testWindupRules/'><strong>org.jboss.windup.rules.tests.WindupRulesTest.testWindupRules</strong></a></li></ul>

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-pr-builder/1362/<h2>Failed Tests: <span class='status-failure'>8</span></h2><h3><a name='windup-pr-builder/org.jboss.windup.rules.apps:windup-rules-xml-tests' /><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1362/org.jboss.windup.rules.apps$windup-rules-xml-tests/testReport'>windup-pr-builder/org.jboss.windup.rules.apps:windup-rules-xml-tests</a>: <span class='status-failure'>8</span></h3><ul><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1362/org.jboss.windup.rules.apps$windup-rules-xml-tests/testReport/org.jboss.windup.rules.apps.xml.rules/ValidateXmlFilesRuleProviderWithInternetTest/testNotValidXml/'><strong>org.jboss.windup.rules.apps.xml.rules.ValidateXmlFilesRuleProviderWithInternetTest.testNotValidXml</strong></a></li><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1362/org.jboss.windup.rules.apps$windup-rules-xml-tests/testReport/org.jboss.windup.rules.apps.xml.rules/ValidateXmlFilesRuleProviderWithInternetTest/testNotValidXsdUrlAttribute/'><strong>org.jboss.windup.rules.apps.xml.rules.ValidateXmlFilesRuleProviderWithInternetTest.testNotValidXsdUrlAttribute</strong></a></li><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1362/org.jboss.windup.rules.apps$windup-rules-xml-tests/testReport/org.jboss.windup.rules.apps.xml.rules/ValidateXmlFilesRuleProviderWithInternetTest/testValidXmlWithoutClassification/'><strong>org.jboss.windup.rules.apps.xml.rules.ValidateXmlFilesRuleProviderWithInternetTest.testValidXmlWithoutClassification</strong></a></li><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1362/org.jboss.windup.rules.apps$windup-rules-xml-tests/testReport/org.jboss.windup.rules.apps.xml.rules/ValidateXmlFilesRuleProviderWithInternetTest/testWithoutXsdUrlAttribute/'><strong>org.jboss.windup.rules.apps.xml.rules.ValidateXmlFilesRuleProviderWithInternetTest.testWithoutXsdUrlAttribute</strong></a></li><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1362/org.jboss.windup.rules.apps$windup-rules-xml-tests/testReport/org.jboss.windup.rules.apps.xml.rules/ValidateXmlFilesRuleProviderWithoutInternetTest/testNotValidXml/'><strong>org.jboss.windup.rules.apps.xml.rules.ValidateXmlFilesRuleProviderWithoutInternetTest.testNotValidXml</strong></a></li><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1362/org.jboss.windup.rules.apps$windup-rules-xml-tests/testReport/org.jboss.windup.rules.apps.xml.rules/ValidateXmlFilesRuleProviderWithoutInternetTest/testNotValidXsdUrlAttribute/'><strong>org.jboss.windup.rules.apps.xml.rules.ValidateXmlFilesRuleProviderWithoutInternetTest.testNotValidXsdUrlAttribute</strong></a></li><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1362/org.jboss.windup.rules.apps$windup-rules-xml-tests/testReport/org.jboss.windup.rules.apps.xml.rules/ValidateXmlFilesRuleProviderWithoutInternetTest/testValidXmlWithoutClassification/'><strong>org.jboss.windup.rules.apps.xml.rules.ValidateXmlFilesRuleProviderWithoutInternetTest.testValidXmlWithoutClassification</strong></a></li><li><a href='http://wonka.mw.lab.eng.bos.redhat.com/jenkins/job/windup-pr-builder/1362/org.jboss.windup.rules.apps$windup-rules-xml-tests/testReport/org.jboss.windup.rules.apps.xml.rules/ValidateXmlFilesRuleProviderWithoutInternetTest/testWithoutXsdUrlAttribute/'><strong>org.jboss.windup.rules.apps.xml.rules.ValidateXmlFilesRuleProviderWithoutInternetTest.testWithoutXsdUrlAttribute</strong></a></li></ul>

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-core-PR-dependents-check/734/

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-pr-builder/1363/

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-core-PR-dependents-check/735/

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-pr-builder/1364/

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-core-PR-dependents-check/736/

File: rules-xml/tests/pom.xml [hunk 1]
```diff
@@ -21,6 +21,18 @@
             <scope>provided</scope>
         </dependency>
 
+        <dependency>
+            <groupId>org.jboss.windup.rules.apps</groupId>
+            <artifactId>windup-rules-xml-impl</artifactId>
+        </dependency>
```
Comment (mbriskar) [inline rules-xml/tests/pom.xml]: is this OK that tests module has dependency on impl module?

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-pr-builder/1365/

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-core-PR-dependents-check/737/

**Comment by mbriskar** [comment]
it depends on impl because I am using the ruleProvider and running it and it is placed in impl module.

**Comment by jsight** [comment]
I think that depending on -impl is ok in that case.

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-pr-builder/1367/

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-core-PR-dependents-check/739/

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-core-PR-dependents-check/740/

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-pr-builder/1368/

File: rules-java-ee/tests/src/test/java/org/jboss/windup/rules/apps/javaee/tests/GenerateJBossEjbDescriptorTest.java [hunk 2]
```diff
@@ -85,18 +86,8 @@ private void checkWebXmls(GraphContext context)
         GraphService<EjbDeploymentDescriptorModel> ejbDescriptors = new GraphService<>(context,EjbDeploymentDescriptorModel.class);
         for (EjbDeploymentDescriptorModel ejbDesc : ejbDescriptors.findAll())
         {
-            Assert.assertTrue(1 >= getIterableSize(ejbDesc.getLinksToTransformedFiles()));
+            Assert.assertTrue(1 >= Iterables.size(ejbDesc.getLinksToTransformedFiles()));
```
Comment (jsight) [inline rules-java-ee/tests/src/test/java/org/jboss/windup/rules/apps/javaee/tests/GenerateJBossEjbDescriptorTest.java]: Good. :)

File: rules-xml/impl/src/main/java/org/jboss/windup/rules/apps/xml/xml/ProcessingIsOnlineGraphCondition.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import org.jboss.windup.config.GraphRewrite;
+import org.jboss.windup.config.condition.GraphCondition;
+import org.jboss.windup.graph.model.WindupConfigurationModel;
+import org.jboss.windup.graph.service.GraphService;
+import org.ocpsoft.rewrite.context.EvaluationContext;
+
+/**
+ * A {@link GraphCondition} returning true/false based on the fact if the current processing is online/offline.
+ */
+public class ProcessingIsOnlineGraphCondition extends GraphCondition
```
Comment (jsight) [inline rules-xml/impl/src/main/java/org/jboss/windup/rules/apps/xml/xml/ProcessingIsOnlineGraphCondition.java]: Good idea... I really thought we already had this. :)

I would suggest moving it to rules-base, though.

File: rules-xml/impl/src/main/java/org/jboss/windup/rules/apps/xml/xml/ValidateXmlFilesRuleProvider.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                    ClassificationService classificationService = new ClassificationService(event.getGraphContext());
+                    ClassificationModel model =classificationService.attachClassification(context, sourceFile, XmlFileModel.NOT_VALID_XML,
+                                null);
+                    TagSetService tagSetService = new TagSetService(event.getGraphContext());
+                    model.setTagModel(tagSetService.getOrCreate(event, Collections.singleton(NOT_VALID_XML_TAG)));
+                }
+            }
+            catch (MalformedURLException e)
+            {
+                e.printStackTrace();
```
Comment (jsight) [inline rules-xml/impl/src/main/java/org/jboss/windup/rules/apps/xml/xml/ValidateXmlFilesRuleProvider.java]: Shouldn't this add a hint as well?

File: rules-xml/tests/pom.xml (hunk 1 above)
Comment (jsight) [inline rules-xml/tests/pom.xml]: I don't think it is a big problem if it is needed for some test.

File: utils/src/main/java/org/jboss/windup/util/Iterables.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package org.jboss.windup.util;
+
+/**
+ * A windup specific class with some useful methods for iterables
+ */
+public class Iterables
```
Comment (jsight) [inline utils/src/main/java/org/jboss/windup/util/Iterables.java]: Why reproduce this here? I believe that we already use guava Iterables.size() in a few places.

**Comment by jsight** [comment]
I have added more questions and comments. The tests look really nice, btw.

File: utils/src/main/java/org/jboss/windup/util/Iterables.java (hunk 5 above)
Comment (mbriskar) [inline utils/src/main/java/org/jboss/windup/util/Iterables.java]: Yes, we do and it caused problems with versioning :-). Should I remove that?

File: rules-xml/impl/src/main/java/org/jboss/windup/rules/apps/xml/xml/ValidateXmlFilesRuleProvider.java (hunk 4 above)
Comment (mbriskar) [inline rules-xml/impl/src/main/java/org/jboss/windup/rules/apps/xml/xml/ValidateXmlFilesRuleProvider.java]: maybe, however I was not able to reproduce this exception so I am not sure when it happens.. I may look into it more

File: rules-xml/impl/src/main/java/org/jboss/windup/rules/apps/xml/xml/ProcessingIsOnlineGraphCondition.java (hunk 3 above)
Comment (mbriskar) [inline rules-xml/impl/src/main/java/org/jboss/windup/rules/apps/xml/xml/ProcessingIsOnlineGraphCondition.java]: good point.

File: utils/src/main/java/org/jboss/windup/util/Iterables.java (hunk 5 above)
Comment (jsight) [inline utils/src/main/java/org/jboss/windup/util/Iterables.java]: Yes, I'd rather just use the guava one for now.

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-core-PR-dependents-check/753/

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-pr-builder/1381/

**Comment by jsight** [comment]
Merged https://github.com/windup/windup/commit/987464af0d80d264860d0f048e9582116f52e88e

**Pull request closed without merging**